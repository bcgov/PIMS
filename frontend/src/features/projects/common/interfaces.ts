import { getCurrentFiscalYear } from '../../../utils/utils';
import { IBuilding, IAddress, IEvaluation, IFiscal } from 'actions/parcelsActions';
import { NoteTypes } from '../../../constants';

/**
 * IProperty interface represents the model used for searching properties.
 */
export interface IProperty {
  id: number;
  projectPropertyId?: number;
  propertyTypeId: number;
  propertyType: string;
  pid: string;
  pin?: number;
  classificationId: number;
  classification: string;
  name: string;
  description: string;
  projectNumbers?: string[];
  latitude: number;
  longitude: number;
  isSensitive: boolean;
  agencyId: number;
  agency: string;
  agencyCode: string;
  subAgency?: string;
  subAgencyCode?: string;

  addressId: number;
  address: string;
  administrativeArea: string;
  province: string;
  postal: string;

  // Financial Values
  market: number;
  marketFiscalYear?: number;
  marketRowVersion?: string;
  netBook: number;
  netBookFiscalYear?: number;
  netBookRowVersion?: string;

  assessedLand?: number;
  assessedLandDate?: Date | string;
  assessedLandFirm?: string;
  assessedLandRowVersion?: string;
  assessedBuilding?: number;
  assessedBuildingDate?: Date | string;
  assessedBuildingFirm?: string;
  assessedBuildingRowVersion?: string;

  // Parcel Properties
  landArea: number;
  landLegalDescription: string;
  zoning?: string;
  zoningPotential?: string;

  // Building Properties
  parcelId?: number;
  constructionTypeId?: number;
  constructionType?: string;
  predominateUseId?: number;
  predominateUse?: string;
  occupantTypeId?: number;
  occupantType?: string;
  floorCount?: number;
  tenancy?: string;
  occupantName?: string;
  leaseExpiry?: Date | string;
  transferLeaseOnSale?: boolean;
  rentableArea?: number;
  rowVersion?: string;
}

/**
 * IPropertyFilter interface, provides a model for querying the API for properties.
 */
export interface IPropertyFilter {
  page: number;
  quantity: number;
  pid?: string;
  address?: string;
  administrativeArea?: string;
  projectNumber?: string;
  ignorePropertiesInProjects?: boolean;
  inSurplusPropertyProgram?: boolean;
  classificationId?: number;
  agencies?: number | number[];
  minLandArea?: number;
  maxLandArea?: number;
  minLotArea?: number;
  maxLotArea?: number;
  all?: boolean;
}

export interface ITask {
  taskId: number;
  name: string;
  description: string;
  sortOrder: number;
  taskType: number;
}

export interface IProjectTask extends ITask {
  projectNumber: number;
  taskId: number;
  taskType: number;
  isCompleted: boolean;
  name: string;
  description: string;
  isOptional: boolean;
  sortOrder: number;
  completedOn: Date;
  statusId: number;
  statusCode: string;
}

export interface IProject {
  id?: number;
  projectNumber: string;
  name: string;
  description: string;
  fiscalYear: number;
  properties: IProperty[];
  projectAgencyResponses: IProjectAgencyResponse[];
  note: string;
  notes: IProjectNote[];
  publicNote: string;
  privateNote: string;
  offersNote?: string;
  agencyId: number;
  agency?: string;
  subAgency?: string;
  statusId: number;
  status?: IStatus;
  exemptionRequested?: boolean;
  exemptionRationale?: string;
  exemptionApprovedOn?: Date | string;
  statusCode?: string;
  tierLevelId: number;
  tasks: IProjectTask[];
  rowVersion?: string;
  confirmation?: boolean;
  approvedOn?: Date | string;
  deniedOn?: Date | string;
  cancelledOn?: Date | string;
  submittedOn?: Date | string;
  documentationNote?: string;
  initialNotificationSentOn?: Date | string;
  thirtyDayNotificationSentOn?: Date | string;
  sixtyDayNoficationSentOn?: Date | string;
  ninetyDayNotificationSentOn?: Date | string;
  onHoldNotificationSentOn?: Date | string;
  interestedReceivedOn?: Date | string;
  interestFromEnhancedReferralNote?: string;
  transferredWithinGreOn?: Date | string;
  clearanceNotificationSentOn?: Date | string;
  requestForSplReceivedOn?: Date | string;
  approvedForSplOn?: Date | string;
  marketedOn?: Date | string;
  disposedOn?: Date | string;
  assessedOn?: Date | string;
  adjustedOn?: Date | string;
  offerAcceptedOn?: Date | string;
  preliminaryFormSignedOn?: Date | string;
  finalFormSignedOn?: Date | string;
  netBook?: number | '';
  assessed?: number | '';
  appraised?: number | '';
  market?: number | '';
  workflowCode?: string;
  purchaser?: string;
  manager?: string;
  actualFiscalYear?: string;
  plannedFutureUse?: string;
  preliminaryFormSignedBy?: string;
  finalFormSignedBy?: string;
  offerAmount?: number | '';
  interestComponent?: number | '';
  loanTermsNote?: string;
  ocgFinancialStatement?: number | '';
  salesCost?: number | '';
  gainBeforeSpl?: number | '';
  netProceeds?: number | '';
  gainNote?: string;
  programCost?: number | '';
  programCostNote?: string;
  priorYearAdjustmentAmount?: number | '';
  remediationNote?: string;
  adjustmentNote?: string;
  closeOutNote?: string;
  salesHistoryNote?: string;
  comments?: string;
  removalFromSplRequestOn?: Date | string;
  removalFromSplApprovedOn?: Date | string;
  removalFromSplRationale?: string;
}

export enum DisposalWorkflows {
  Erp = 'ERP',
  Spl = 'SPL',
}

export enum DisposeWorkflowStatus {
  Draft = 'DR',
  SelectProperties = 'DR-P',
  UpdateInformation = 'DR-I',
  RequiredDocumentation = 'DR-D',
  Approval = 'DR-A',
  Review = 'DR-RE',
}

export enum ReviewWorkflowStatus {
  PropertyReview = 'AS-I',
  DocumentReview = 'AS-D',
  AppraisalReview = 'AS-AP',
  FirstNationConsultation = 'AS-FNC',
  ExemptionReview = 'AS-EXE',
  ExemptionProcess = 'AS-EXP',
  ApprovedForErp = 'AP-ERP',
  OnHold = 'ERP-OH',
  ERP = 'ERP-ON',
  TransferredGRE = 'T-GRE',
  ApprovedForSpl = 'AP-SPL',
  PreMarketing = 'SPL-PM',
  OnMarket = 'SPL-M',
  ContractInPlaceConditional = 'SPL-CIP-C',
  ContractInPlaceUnconditional = 'SPL-CIP-U',
  NotInSpl = 'AP-!SPL',
  ApprovedForExemption = 'AP-EXE',
  Denied = 'DE',
  Disposed = 'DIS',
  Cancelled = 'CA',
  InErp = 'ERP-ON',
}

export enum SPPApprovalTabs {
  projectInformation = 'projectInformation',
  documentation = 'documentation',
  erp = 'enhancedReferralProcess',
  spl = 'surplusPropertyList',
  closeOutForm = 'closeOutForm',
}

export interface ProjectWorkflowComponent {
  component: React.ComponentType<any>;
  workflowStatus: DisposeWorkflowStatus;
}

export interface IStatus {
  id: number;
  name: string;
  sortOrder: number;
  description: string;
  route: string;
  workflowCode: string;
  code: string;
  isMilestone: boolean;
  tasks: IProjectTask[];
  isOptional: boolean;
  toStatus?: IStatus[];
  isActive: boolean;
  parentId?: number;
}

export interface IStepProps {
  isReadOnly?: boolean;
  formikRef?: any;
}

export interface IProjectProperty {
  id?: number;
  projectNumber?: string;
  propertyType: string;
  parcelId?: number;
  parcel?: IApiProperty;
  buildingId?: number;
  building?: IApiProperty;
}

export const initialValues: IProject = {
  projectNumber: '',
  name: '',
  note: '',
  notes: [],
  description: '',
  properties: [],
  tierLevelId: 1,
  statusId: 1,
  agencyId: 0,
  tasks: [],
  projectAgencyResponses: [],
  publicNote: '',
  privateNote: '',
  statusCode: 'DR',
  fiscalYear: getCurrentFiscalYear(),
  confirmation: false,
  approvedOn: '',
  deniedOn: '',
  cancelledOn: '',
  submittedOn: '',
  exemptionRequested: false,
  exemptionRationale: '',
  exemptionApprovedOn: '',
  initialNotificationSentOn: '',
  thirtyDayNotificationSentOn: '',
  sixtyDayNoficationSentOn: '',
  ninetyDayNotificationSentOn: '',
  onHoldNotificationSentOn: '',
  interestedReceivedOn: '',
  interestFromEnhancedReferralNote: '',
  transferredWithinGreOn: '',
  clearanceNotificationSentOn: '',
  requestForSplReceivedOn: '',
  approvedForSplOn: '',
  marketedOn: '',
  disposedOn: '',
  offerAcceptedOn: '',
  assessedOn: '',
  adjustedOn: '',
  offersNote: '',
  preliminaryFormSignedOn: '',
  finalFormSignedOn: '',
  netBook: '',
  assessed: '',
  appraised: '',
  market: '',
  workflowCode: '',
  offerAmount: '',
  purchaser: '',
  manager: '',
  actualFiscalYear: '',
  plannedFutureUse: '',
  preliminaryFormSignedBy: '',
  finalFormSignedBy: '',
  interestComponent: '',
  ocgFinancialStatement: '',
  salesCost: '',
  gainBeforeSpl: '',
  netProceeds: '',
  programCost: '',
  priorYearAdjustmentAmount: '',
  removalFromSplRequestOn: '',
  removalFromSplApprovedOn: '',
  removalFromSplRationale: '',
  documentationNote: '',
  salesHistoryNote: '',
  comments: '',
  loanTermsNote: '',
  closeOutNote: '',
  remediationNote: '',
  programCostNote: '',
  gainNote: '',
  adjustmentNote: '',
};

export interface IApiProject {
  id?: number;
  projectNumber: string;
  fiscalYear: number;
  name: string;
  description: string;
  properties: IProjectProperty[];
  projectAgencyResponses: IProjectAgencyResponse[];
  note: string;
  netBook: number | '';
  market: number | '';
  assessed: number | '';
  appraised?: number | '';
  publicNote: string;
  privateNote: string;
  exemptionRequested?: boolean;
  exemptionRationale?: string;
  exemptionApprovedOn?: Date | string;
  interestFromEnhancedReferralNote?: string;
  agencyId: number;
  statusId: number;
  statusCode?: string;
  tierLevelId: number;
  tasks: IProjectTask[];
  rowVersion?: string;
  approvedOn?: Date | string;
  deniedOn?: Date | string;
  cancelledOn?: Date | string;
  submittedOn?: Date | string;
  InitialNotificationSentOn?: Date | string;
  ThirtyDayNotificationSentOn?: Date | string;
  SixtyDayNoficationSentOn?: Date | string;
  NinetyDayNotificationSentOn?: Date | string;
  OnHoldNotificationSentOn?: Date | string;
  TransferredWithinGreOn?: Date | string;
  ClearanceNotificationSentOn?: Date | string;
  disposedOn?: Date | string;
  notes: IProjectNote[];
}

export interface IProjectNote {
  id?: number;
  noteType: string | NoteTypes;
  note?: string;
  rowVersion?: string;
  projectId?: number;
}

export interface IApiProperty {
  id: number;
  parcelId?: number;
  buildingId?: number;
  pid?: string;
  pin?: number | '';
  projectNumbers: string[];
  latitude: number;
  longitude: number;
  classification?: string;
  classificationId: number;
  name: string;
  description: string;
  address?: IAddress;
  landArea: number;
  landLegalDescription: string;
  zoning: string;
  zoningPotential: string;
  agency?: string;
  subAgency?: string;
  agencyId: number;
  isSensitive: boolean;
  buildings: IBuilding[];
  evaluations: IEvaluation[];
  fiscals: IFiscal[];
  rowVersion?: string;
}

export enum AgencyResponses {
  Ignore = 'Ignore',
  Watch = 'Watch',
}

export interface IProjectAgencyResponse {
  response: AgencyResponses;
  notificationId?: number;
  agencyId: number;
  agencyCode?: string;
  projectId: number;
  receivedOn?: string;
  note?: string;
  rowVersion?: string;
  offerAmount?: number;
}
