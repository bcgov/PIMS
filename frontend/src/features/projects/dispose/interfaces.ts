import { IBuilding, IAddress, IEvaluation, IFiscal } from 'actions/parcelsActions';

/**
 * IProperty interface represents the model used for searching properties.
 */
export interface IProperty {
  id: number;
  propertyTypeId: number;
  propertyType: string;
  pid: string;
  pin?: number;
  statusId: number;
  status?: string;
  classificationId: number;
  classification: string;
  description: string;
  projectNumber?: string;
  latitude: number;
  longitude: number;
  isSensitive: boolean;
  agencyId: number;
  agency: string;
  agencyCode: string;
  subAgency?: string;
  subAgencyCode?: string;

  addressId: number;
  address: string;
  cityId: number;
  city: string;
  province: string;
  postal: string;

  // Financial Values
  estimated: number;
  estimatedFiscalYear?: number;
  estimatedRowVersion?: string;
  netBook: number;
  netBookFiscalYear?: number;
  netBookRowVersion?: string;

  assessed: number;
  assessedDate?: Date | string;
  assessedRowVersion?: string;
  appraised: number;
  appraisedDate?: Date | string;
  appraisedRowVersion?: string;

  // Parcel Properties
  landArea: number;
  landLegalDescription: string;
  municipality: string;
  zoning: string;
  zoningPotential: string;

  // Building Properties
  parcelId?: number;
  localId?: string;
  constructionTypeId?: number;
  constructionType?: string;
  predominateUseId?: number;
  predominateUse?: string;
  occupantTypeId?: number;
  occupantType?: string;
  floorCount?: number;
  tenancy?: string;
  occupantName?: string;
  leaseExpiry?: Date | string;
  transferLeaseOnSale?: boolean;
  rentableArea?: number;
  rowVersion?: string;
}

/**
 * IPropertyFilter interface, provides a model for querying the API for properties.
 */
export interface IPropertyFilter {
  page: number;
  quantity: number;
  address?: string;
  municipality?: string;
  projectNumber?: string;
  ignorePropertiesInProjects?: boolean;
  classificationId?: number;
  statusId?: number;
  agencies?: number | number[];
  minLandArea?: number;
  maxLandArea?: number;
  minLotArea?: number;
  maxLotArea?: number;
  all?: boolean;
}
export interface IProjectTask {
  projectNumber: number;
  taskId: number;
  taskType: string;
  isCompleted: boolean;
  name: string;
  description: string;
  isOptional: boolean;
  sortOrder: number;
  completedOn: Date;
}
export interface IProject {
  projectNumber: string;
  name: string;
  description: string;
  properties: IProperty[];
  note: string;
  agencyId: number;
  statusId: number;
  tierLevelId: number;
  tasks: IProjectTask[];
  rowVersion?: string;
  confirmation?: boolean;
}

export enum DisposeWorkflowStatus {
  Draft = 0,
  SelectProperties,
  UpdateInformation,
  RequiredDocumentation,
  Approval,
  Submitted,
}

export interface ProjectWorkflowComponent {
  component: React.ComponentType<any>;
  workflowStatus: DisposeWorkflowStatus;
}

export interface IStepProps {
  isReadOnly?: boolean;
  formikRef?: any;
}

export interface IProjectProperty {
  id?: number;
  projectNumber: string;
  propertyType: string;
  parcelId?: number;
  parcel?: IApiProperty;
  buildingId?: number;
  building?: IApiProperty;
}

export const initialValues: IProject = {
  projectNumber: '',
  name: '',
  note: '',
  description: '',
  properties: [],
  tierLevelId: 1,
  statusId: 0,
  agencyId: 0,
  tasks: [],
};

export interface IApiProject {
  projectNumber: string;
  name: string;
  description: string;
  projectProperties: IProjectProperty[];
  note: string;
  agencyId: number;
  statusId: number;
  tierLevelId: number;
  tasks: IProjectTask[];
  rowVersion?: string;
}

export interface IApiProperty {
  id: number;
  parcelId?: number;
  buildingId?: number;
  pid?: string;
  pin?: number | '';
  projectNumber: string;
  latitude: number;
  longitude: number;
  statusId: number;
  propertyStatus?: string;
  classification?: string;
  classificationId: number;
  description: string;
  address?: IAddress;
  landArea: number;
  landLegalDescription: string;
  zoning: string;
  zoningPotential: string;
  municipality: string;
  agency?: string;
  subAgency?: string;
  agencyId: number;
  isSensitive: boolean;
  buildings: IBuilding[];
  evaluations: IEvaluation[];
  fiscals: IFiscal[];
  rowVersion?: string;
}
