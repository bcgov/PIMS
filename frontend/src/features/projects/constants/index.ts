export * from './AgencyResponses';
export * from './DisposalWorkflows';
export * from './DisposeWorkflowStatus';
export * from './ProjectWorkflowComponents';
export * from './ReviewWorkflowStatus';
export * from './SPPApprovalTabs';
