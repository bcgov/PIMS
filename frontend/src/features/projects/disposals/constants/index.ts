export * from './defaultFilter';
export * from './defaultProjectForm';
