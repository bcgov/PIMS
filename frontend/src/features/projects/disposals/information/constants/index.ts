export * from './AddPropertyColumns';
export * from './PropertyColumns';
