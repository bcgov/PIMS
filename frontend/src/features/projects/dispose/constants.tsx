import { DisposeWorkflowStatus, ProjectWorkflowComponent } from '.';
import React from 'react';

//TODO: map to real components
export const projectWorkflowComponents: ProjectWorkflowComponent[] = [
  { component: () => <p>Step 1</p>, workflowStatus: DisposeWorkflowStatus.Draft },
  { component: () => <p>Step 2</p>, workflowStatus: DisposeWorkflowStatus.SelectProperties },
  { component: () => <p>Step 3</p>, workflowStatus: DisposeWorkflowStatus.UpdateInformation },
  { component: () => <p>Step 4</p>, workflowStatus: DisposeWorkflowStatus.RequiredDocumentation },
  { component: () => <p>Step 5</p>, workflowStatus: DisposeWorkflowStatus.Approval },
  { component: () => <p>Step 6</p>, workflowStatus: DisposeWorkflowStatus.Submitted },
];
