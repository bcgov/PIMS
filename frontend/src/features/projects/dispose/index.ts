export * from './interfaces';
export * from './ProjectWorkflowComponents';
export * from './components/FilterBar';
export { default as FilterBar } from './components/FilterBar';
export { default as ProjectDisposeView } from './ProjectDisposeView';
export * from './components/StepActions';
export { default as StepStatusIcon } from './components/StepStatusIcon';
export * from './components/ReviewApproveActions';
export { default as ProjectNotes } from './components/ProjectNotes';
export { default as ExemptionRequest } from './components/ExemptionRequest';
export { default as EditButton } from './components/EditButton';
export { default as SelectProjectPropertiesPage } from './components/SelectProjectPropertiesPage';
export { ApprovalActions } from './components/ApprovalActions';

export { default as projectSlice } from './slices/projectSlice';
export { default as projectTasksSlice } from './slices/projectTasksSlice';
export { default as projectWorflowSlice } from './slices/projectWorkflowSlice';

export { default as ApprovalConfirmationStep } from './steps/ApprovalConfirmationStep';
export { default as SelectProjectPropertiesStep } from './steps/SelectProjectPropertiesStep';
export { default as UpdateInfoStep } from './steps/UpdateInfoStep';
export { default as DocumentationStep } from './steps/DocumentationStep';
export { default as ProjectDraftStep } from './steps/ProjectDraftStep';
export { default as ReviewProjectStep } from './steps/ReviewProjectStep';
export { default as StepErrorSummary } from './steps/StepErrorSummary';
export { default as ReviewApproveStep } from './steps/ReviewApproveStep';
export { default as ApprovalStep } from './steps/ApprovalStep';

export { default as ApprovalConfirmationForm } from './forms/ApprovalConfirmationForm';
export { default as UpdateInfoForm } from './forms/UpdateInfoForm';
export { default as DocumentationForm } from './forms/DocumentationForm';
export { default as ProjectDraftForm } from './forms/ProjectDraftForm';
export { default as AgencyResponseForm } from './forms/AgencyResponseForm';
export { default as ReviewProjectForm } from './forms/ReviewProjectForm';
export { default as ReviewApproveForm } from './forms/ReviewApproveForm';
export { default as AppraisalCheckListForm } from './forms/AppraisalCheckListForm';
export { default as FirstNationsCheckListForm } from './forms/FirstNationsCheckListForm';
export { default as SelectProjectPropertiesForm } from './forms/SelectProjectPropertiesForm';
export { default as ApprovalForm } from './forms/ApprovalForm';
export { default as EnhancedReferralCompleteForm } from './forms/EnhancedReferralCompleteForm';

export { default as ProjectInformationTab } from './tabs/ProjectInformationTab';
export { default as DocumentationTab } from './tabs/DocumentationTab';
export { default as EnhancedReferralTab } from './tabs/EnhancedReferralTab';

export { default as useStepper } from './hooks/useStepper';
export { default as useStepForm } from './hooks/useStepForm';
export * from './hooks/stepperContext';

export * from './slices/projectSlice';
export * from './slices/projectTasksSlice';
export * from './slices/projectWorkflowSlice';

export * from './forms/disposalYupSchema';
export * from './strings';
