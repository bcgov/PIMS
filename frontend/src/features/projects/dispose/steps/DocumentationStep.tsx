import { Form } from 'components/common/form';
import { DisposeWorkflowStatus } from 'features/projects/constants';
import { IProject, IProjectTask, IStepProps } from 'features/projects/interfaces';
import { Formik, setIn, yupToFormErrors } from 'formik';
import _ from 'lodash';
import React from 'react';
import { Container } from 'react-bootstrap';

import { DocumentationForm, ProjectNotes, StepErrorSummary, useStepForm } from '../../common';
import { EnhancedReferralExemptionSchema, useStepper } from '..';

const handleValidate = (project: IProject) => {
  return project.tasks.reduce((errors: any, task: IProjectTask, index: number) => {
    if (
      !task.isCompleted &&
      !task.isOptional &&
      task.statusCode === DisposeWorkflowStatus.RequiredDocumentation
    ) {
      errors = setIn(errors, `tasks.${index}.isCompleted`, 'Required');
    }
    try {
      EnhancedReferralExemptionSchema.validateSync(project, {
        abortEarly: false,
      });
    } catch (schemaErrors) {
      return _.merge(errors, yupToFormErrors(schemaErrors));
    }

    return errors;
  }, {});
};

/**
 * Displays all tasks from TASK table as clickable checkboxes.
 * @param param0 {isReadOnly formikRef} formikRef allow remote formik access, isReadOnly toggle to prevent updates.
 */
const DocumentationStep = ({ isReadOnly, formikRef }: IStepProps) => {
  const { onSubmit, canUserEditForm } = useStepForm();
  const { project } = useStepper();
  const tasks = _.filter(project.tasks, {
    statusCode: DisposeWorkflowStatus.RequiredDocumentation,
  });
  if (!project) {
    // Step does not allow creation of new properties
    throw Error('Unexpected error updating project. Please reload your project.');
  }

  return (
    <Container fluid className="DocumentationStep">
      <Formik
        initialValues={project}
        innerRef={formikRef}
        validate={handleValidate}
        validateOnChange={false}
        enableReinitialize={true}
        onSubmit={(values, actions) => {
          // set the completed on dates during submission.
          values.tasks.forEach((task: any, index: any) => {
            const existingTask = project.tasks[index];

            //if any tasks have been completed for the first time set the completed date.
            if (task.isCompleted && !existingTask.isCompleted) {
              task.completedOn = new Date();
            }
          });
          return onSubmit(values, actions);
        }}
      >
        {() => (
          <Form>
            <DocumentationForm
              isReadOnly={isReadOnly || !canUserEditForm(project.agencyId)}
              tasks={tasks}
            />
            {!isReadOnly && <ProjectNotes className="col-md-auto" />}
            <StepErrorSummary />
          </Form>
        )}
      </Formik>
    </Container>
  );
};

export default DocumentationStep;
