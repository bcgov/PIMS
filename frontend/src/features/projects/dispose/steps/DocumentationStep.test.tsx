import { useKeycloak } from '@react-keycloak/web';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { ProjectActions } from 'constants/actionTypes';
import * as reducerTypes from 'constants/reducerTypes';
import { DisposeWorkflowStatus } from 'features/projects/constants';
import { IProjectTask } from 'features/projects/interfaces';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import DocumentationStep from './DocumentationStep';

const mockAxios = new MockAdapter(axios);
mockAxios.onAny().reply(200, {});
jest.mock('@react-keycloak/web');
(useKeycloak as jest.Mock).mockReturnValue({
  keycloak: {
    userInfo: {
      agencies: [1],
      roles: [],
    },
    subject: 'test',
  },
});

const mockStore = configureMockStore([thunk]);
const history = createMemoryHistory();

const tasks: IProjectTask[] = [
  {
    projectNumber: 1,
    isCompleted: false,
    isOptional: false,
    completedOn: new Date(),
    taskId: 1,
    name: 'task-0',
    sortOrder: 0,
    description: 'Task #1',
    taskType: 1,
    statusId: 0,
    statusCode: DisposeWorkflowStatus.RequiredDocumentation,
  },
  {
    projectNumber: 1,
    isCompleted: false,
    isOptional: false,
    completedOn: new Date(),
    taskId: 2,
    name: 'task-1',
    sortOrder: 0,
    description: 'Task #2',
    taskType: 1,
    statusId: 0,
    statusCode: DisposeWorkflowStatus.RequiredDocumentation,
  },
];

const store = mockStore({
  [reducerTypes.ProjectReducers.PROJECT]: { project: { tasks: tasks } },
  [reducerTypes.ProjectReducers.TASKS]: tasks,
  [reducerTypes.NETWORK]: {
    requests: { [ProjectActions.GET_PROJECT]: {} },
  },
});

const uiElement = (
  <Provider store={store}>
    <Router history={history}>
      <DocumentationStep />
    </Router>
  </Provider>
);

describe('Documentation Step', () => {
  afterEach(() => {
    cleanup();
  });
  it('renders correctly', () => {
    const { container } = render(uiElement);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correct labels', () => {
    const { getByText } = render(uiElement);
    expect(getByText('Task #1')).toBeInTheDocument();
    expect(getByText('Task #2')).toBeInTheDocument();
  });

  it('documentation validation works', async () => {
    const { getAllByText, container } = render(uiElement);
    const form = container.querySelector('form');
    await waitFor(() => {
      fireEvent.submit(form!);
    });
    expect(getAllByText('Required')).toHaveLength(2);
  });
});
