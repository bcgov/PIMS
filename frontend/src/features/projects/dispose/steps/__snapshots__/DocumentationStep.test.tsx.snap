// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documentation Step renders correctly 1`] = `
<div
  className="DocumentationStep container-fluid"
>
  <form
    className=""
    noValidate={true}
    onSubmit={[Function]}
  >
    <h3>
      Documentation
    </h3>
    <div
      className="tasksForm form-row"
    >
      <div
        className="required form-group"
      >
        <div
          className="check-field"
        >
          <div
            className="form-check"
          >
            <input
              checked={false}
              className="form-check-input position-static"
              disabled={true}
              id="input-tasks.0.isCompleted"
              name="tasks.0.isCompleted"
              onChange={[Function]}
              required={true}
              type="checkbox"
              value={false}
            />
          </div>
          <label
            className="form-label"
            htmlFor="input-tasks.0.isCompleted"
          >
            Task #1
          </label>
        </div>
      </div>
    </div>
    <div
      className="tasksForm form-row"
    >
      <div
        className="required form-group"
      >
        <div
          className="check-field"
        >
          <div
            className="form-check"
          >
            <input
              checked={false}
              className="form-check-input position-static"
              disabled={true}
              id="input-tasks.1.isCompleted"
              name="tasks.1.isCompleted"
              onChange={[Function]}
              required={true}
              type="checkbox"
              value={false}
            />
          </div>
          <label
            className="form-label"
            htmlFor="input-tasks.1.isCompleted"
          >
            Task #2
          </label>
        </div>
      </div>
    </div>
    <div
      className="ProjectNotes form-row"
    >
      <div
        className="col-md-10 form-group"
      >
        <label
          className="form-label"
          htmlFor="input-note"
        >
          Notes
        </label>
        <textarea
          className="col-md-5 form-control"
          id="input-note"
          name="note"
          onBlur={[Function]}
          onChange={[Function]}
        />
      </div>
    </div>
    <div
      style={
        Object {
          "textAlign": "right",
        }
      }
    />
  </form>
</div>
`;
